Realm/ObjectStore/impl/background_collection.hpp

```diff
@@ -0,0 +1,205 @@
+    // Public API for the collections using this to get notifications:
+
+    // Stop receiving notifications from this background worker
+    // This must be called in the destructor of the collection
+    void unregister() noexcept;
+
+    // Add a callback to be called each time the collection changes
+    // This can only be called from the target collection's thread
+    // Returns a token which can be passed to remove_callback()
+    size_t add_callback(CollectionChangeCallback callback);
+    // Remove a previously added token. The token is no longer valid after
+    // calling this function and must not be used again. This function can be
+    // called from any thread.
+    void remove_callback(size_t token);
+
+    // ------------------------------------------------------------------------
+    // API for RealmCoordinator to manage running things and calling callbacks
+
+    Realm* get_realm() const noexcept { return m_realm.get(); }
+
+    // Get the SharedGroup version which this collection can attach to (if it's
+    // in handover mode), or can deliver to (if it's been handed over to the BG worker alredad)
+    SharedGroup::VersionID version() const noexcept { return m_sg_version; }
+
+    // Release references to all core types
+    // This is called on the worker thread to ensure that non-thread-safe things
+    // can be destroyed on the correct thread, even if the last reference to the
+    // BackgroundCollection is released on a different thread
+    virtual void release_data() noexcept = 0;
+
+    // Call each of the currently registered callbacks, if there have been any
+    // changes since the last time each of those callbacks was called
+    void call_callbacks();
+
+    bool is_alive() const noexcept;
+
+    // Attach the handed-over query to `sg`. Must not be already attached to a SharedGroup.
+    void attach_to(SharedGroup& sg);
+    // Create a new query handover object and stop using the previously attached
+    // SharedGroup
+    void detach();
+
+    // Set `info` as the new ChangeInfo that will be populated by the next
+    // transaction advance, and register all required information in it
+    void add_required_change_info(TransactionChangeInfo& info);
+
+    virtual void run() { }
+    void prepare_handover();
+    bool deliver(SharedGroup&, std::exception_ptr);
+
+protected:
+    bool have_callbacks() const noexcept { return m_have_callbacks; }
+    void add_changes(CollectionChangeBuilder change) { m_accumulated_changes.merge(std::move(change)); }
+    void set_table(Table const& table);
+    std::unique_lock<std::mutex> lock_target();
+
+private:
+    virtual void do_attach_to(SharedGroup&) = 0;
+    virtual void do_detach_from(SharedGroup&) = 0;
+    virtual void do_prepare_handover(SharedGroup&) = 0;
+    virtual bool do_deliver(SharedGroup&) { return true; }
+    virtual bool do_add_required_change_info(TransactionChangeInfo&) { return true; }
```

Comment:
What does the bool return value of `do_deliver` and `do_add_required_change_info` signify?